File: cyanBat/src/main/java/at/msmiech/cyanbat/threads/ObstacleGenerator.java
package at.msmiech.cyanbat.threads;

import java.util.List;
import java.util.Random;

import android.util.Log;

import at.grueneis.game.framework.Pixmap;
import at.msmiech.cyanbat.CyanBatGame;
import at.msmiech.cyanbat.interfaces.GameObject;
import at.msmiech.cyanbat.objects.CollisionDetection;
import at.msmiech.cyanbat.objects.gameobjects.Obstacle;
import at.msmiech.cyanbat.screens.CyanBatBaseScreen;

public class ObstacleGenerator extends Thread {

    private static final boolean DEBUG = false;
    private static final String TAG = CyanBatGame.TAG;
    private static final int OBSTACLE_GENERATION_INTERVAL = 1000; // in ms
    private final List<GameObject> gameObjects;
    private final Random rnd = new Random();
    private CollisionDetection collisionDetection;

    public ObstacleGenerator(List<GameObject> gameObjects) {
        this.gameObjects = gameObjects;
    }

    @Override
    public void run() {
        try {
            while (!interrupted()) {
                sleep(rnd.nextInt(OBSTACLE_GENERATION_INTERVAL) + OBSTACLE_GENERATION_INTERVAL);
                generateObstacle();
            }
        } catch (InterruptedException e) {
            this.interrupt();
        }
        super.run();
    }

    private void generateObstacle() {
        if (DEBUG)
            Log.d(TAG, "generateObstacle");
        int obstacleHeight;
        Pixmap obstaclePixmap;
        if (rnd.nextBoolean()) // Top or not?
        {
            obstaclePixmap = CyanBatGame.topObstacles[rnd
                    .nextInt(CyanBatGame.topObstacles.length)];
            obstacleHeight = 0;

        } else {
            obstaclePixmap = CyanBatGame.bottomObstacles[rnd
                    .nextInt(CyanBatGame.bottomObstacles.length)];
            obstacleHeight = CyanBatBaseScreen.DISPLAY_WIDTH
                    - obstaclePixmap.getHeight();
        }
        synchronized (gameObjects) {
            Obstacle obs = new Obstacle(CyanBatBaseScreen.DISPLAY_HEIGHT,
                    obstacleHeight, obstaclePixmap);
            gameObjects.add(obs);
            if(collisionDetection != null) {
                collisionDetection.addObjectToCheck(obs);
            }
        }
    }

    public void setCollisionDetection(CollisionDetection collisionDetection) {
        this.collisionDetection = collisionDetection;
    }
}

File: app/src/main/kotlin/at/smiech/cyanbat/service/ObstacleGenerator.kt
package at.smiech.cyanbat.service

import android.util.Log
import at.smiech.cyanbat.gameobject.GameObject
import at.smiech.cyanbat.gameobject.impl.Obstacle
import at.smiech.cyanbat.resource.Level
import at.smiech.cyanbat.util.DEBUG
import at.smiech.cyanbat.util.TAG
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.random.Random
import kotlin.time.Duration.Companion.milliseconds

class ObstacleGenerator(
    private val worldWidth: Int,
    private val worldHeight: Int,
    private val gameObjects: MutableList<GameObject>,
    var level: Level
) {
    private var collisionDetector: CollisionDetector? = null
    private var generationInterval = OBSTACLE_GENERATION_INTERVAL

    private var waitJob: Job? = null

    @OptIn(DelicateCoroutinesApi::class)
    fun generateObstacle() {
        if (waitJob?.isActive == true) {
            return
        }

        if (DEBUG) {
            Log.d(TAG, "generateObstacle")
        }

        waitJob = GlobalScope.launch {
            delay((Random.nextInt(generationInterval) + generationInterval).toLong().milliseconds)
        }

        var y = 0
        val obstaclePixmap = if (Random.nextBoolean())
        // Top or not?
        {
            level.topObstacles[Random.nextInt(level.topObstacles.size)]

        } else {
            level.bottomObstacles[Random.nextInt(level.bottomObstacles.size)]?.also {
                y = worldHeight - it.height
            }
        }

        obstaclePixmap?.let {
            Obstacle(
                worldWidth,
                y, it
            ).also {
                gameObjects.add(it)
                collisionDetector?.addObjectToCheck(it)
            }
        }
    }


    fun setCollisionDetection(collisionDetector: CollisionDetector) {
        this.collisionDetector = collisionDetector
    }

    companion object {
        private const val OBSTACLE_GENERATION_INTERVAL = 1000 // in ms
    }
}
